PROTOCOL.md

```diff
@@ -0,0 +1,127 @@
+payloads. The protocol is line-based, and each measurement (data point)
+must adhere to one of the the following formats:
+
+    <measurement name>:<value>|<type>|#<tag name>=<tag value>[,<tag name>=<tag value>[,...]]<newline>
+
+    <measurement name>:1|c|#<tag name>=<tag value>[,<tag name>=<tag value[,...]]<newline>
+
+* `measurement name` is an arbitrary string
+* `value` is the recorded measurement, if type is other than `c`
+* `type` is one of the following:
+  * `g` - gauge; an already measured value
+  * `ms` - timer (milliseconds); duration of an operation
+  * `h` - histogram; distribution of (often discrete) values
+  * `s` - set; collection of distinct keys
+  * `c` - counter; an automatically increasing count of occurrences
+* `tag name` is an arbitrary label or tag for the value
+* `tag value` is an arbitrary categorisation
+* `newline` is a literal newline
+
+A message can contain zero or more tags, but usually you want to provide
+at least one to assign the measurement context.
+
+The tags are commonly used for grouping when viewing the generated
+time-series graphs, as well as for specifying alert rules.
+
+### Examples
+
+Let's say you want to track the roundrip times, as well as received
+and generated status codes for a login service. You measure the time
+spent to make a call to another service, or maybe to read values from a
+data storage.
+
+For one particular operation the read took 4.1 milliseconds. You have
+chosen to use "duration" as the measurement name for anything involving
+execution time. In that case, the StatsD message payload for the
+roundtrip time taken could be:
+
+  - `duration:4.1|ms|#service=login,team=myteam,operation=read\n`
+
+In order to track the status of their service, the team also records
+each returned status code. For a service with a non-trivial amount of
+activity, it is important to know how many times any particular status
+code has been served. So, just before returning the service response,
+the service could send the following message to bucky3:
+
+  - `status:200|h|#service=login,team=myteam,route=/user/login\n`
+
+Of course, if a login was unsuccesful, the message could instead be:
+
+  - `status:401|h|#service=login,team=myteam,route=/user/login\n`
+
+Another team maintains a service with persistent real-time connections.
+In addition to response status codes, they care roundtrip times and the
+number of connected clients. Every couple of seconds, they record the
+number of currently open client connections, and at a particular moment
+they had 473 connections. The message payload could then be:
+
+  - `connections:473|g|#service=ourstream,team=otherteam\n`
+
+In addition, every time they serve a request, they record both the
+response status and the roundtrip time. So on every succesful response
+they could send the following two messages:
+
+- `status:200|h|#service=ourstream,team=otherteam,action=something\n`
+- `duration:2.9|ms|#service=ourstream,team=otherteam,action=something\n`
```

Comment:
Another option is to integrate status into duration as in `duration:1|ms|#action=something,result=200` - that works ok when the set of all `result` values is small (like in HTTP). For big sets (like login IDs), it is not a good choice though.

Reply:
Hum, true. I'll think about how to express these nicely.